File: ports/java/src/core/tools/BcsVariant.java
package bcs.core.tools;

// BcsVariant acts as a dynamic type wrapper mapping natively to Java's `Object`
public class BcsVariant {
    private Object value;

    public BcsVariant(Object value) {
        this.value = value;
    }

    public Object getValue() {
        return value;
    }

    public void setValue(Object value) {
        this.value = value;
    }

    public boolean isNil() {
        return value == null;
    }

    public String getType() {
        return value != null ? value.getClass().getSimpleName() : "Nil";
    }

    @Override
    public String toString() {
        return value != null ? value.toString() : "Nil";
    }

    public Integer toInt() {
        if (value instanceof Integer) {
            return (Integer) value;
        }
        return null;
    }

    public Double toFloat() {
        if (value instanceof Double) {
            return (Double) value;
        }
        return null;
    }

    public Boolean toBool() {
        if (value instanceof Boolean) {
            return (Boolean) value;
        }
        return null;
    }
}

File: ports/csharp/src/core/tools/BcsVariant.cs
namespace BCS.Core.Tools {
    // BcsVariant acts as a dynamic type wrapper mapping natively to C#'s `object`
    public class BcsVariant {
        public object Value { get; set; }

        public BcsVariant(object value) {
            Value = value;
        }

        public bool IsNil() {
            return Value == null;
        }

        public string GetVariantType() {
            return Value?.GetType().Name ?? "Nil";
        }

        public override string ToString() {
            return Value?.ToString() ?? "Nil";
        }

        public int? ToInt() {
            if (Value is int val) return val;
            return null;
        }

        public double? ToFloat() {
            if (Value is double val) return val;
            return null;
        }

        public bool? ToBool() {
            if (Value is bool val) return val;
            return null;
        }
    }
}
